core/src/main/java/ai/timefold/solver/core/impl/score/director/ValueRangeManager.java

```diff
@@ -358,59 +365,255 @@ public <T> CountableValueRange<T> getFromSolution(ValueRangeDescriptor<Solution_
+            return valueRange;
+        }
+        var leftSorter = (SelectionSorter<Solution_, T>) item.leftSorter();
+        var leftValueRange = (CountableValueRange<T>) item.leftItem();
+        var rightSorter = (SelectionSorter<Solution_, T>) item.rightSorter();
+        var rightValueRange = (CountableValueRange<T>) item.rightItem();
+        if (leftValueRange != null && (sorter == null || Objects.equals(leftSorter, sorter))) {
+            // Return the left value if there is no sorter or if the left sorter is the same as the provided one.
+            return leftValueRange;
+        } else if (rightValueRange != null && Objects.equals(rightSorter, sorter)) {
+            // Return the right value only if the right sorter is the same as the provided one.
+            return rightValueRange;
+        }
+        // If the left sorter is null, and the given sorter is not, we replace the left value with a sorted one.
+        if (leftSorter == null) {
+            if (!(leftValueRange instanceof SortableValueRange sortableValueRange)) {
+                throw new IllegalStateException(
+                        "Impossible state: value range (%s) on planning solution (%s) is not sortable."
+                                .formatted(valueRangeDescriptor, solution));
+            }
+            var sorterAdapter = SelectionSorterAdapter.of(solution, sorter);
+            var valueRange = (CountableValueRange<T>) sortableValueRange.sort(sorterAdapter);
+            fromSolution[valueRangeDescriptor.getOrdinal()] =
+                    new ValueRangeItem<>(null, valueRange, sorter, rightValueRange, rightSorter);
+            // We need to update the index map or the positions may become inconsistent
+            Class<?> valueClass = findValueClass(valueRange);
+            var valueIndexMap = buildIndexMap(valueRange.createOriginalIterator(), (int) valueRange.getSize(),
+                    ConfigUtils.isGenericTypeImmutable(valueClass));
+            fromSolutionValueIndexMap[valueRangeDescriptor.getOrdinal()] = (Map<Object, Integer>) valueIndexMap;
+            return valueRange;
+        } else if (rightValueRange == null) {
+            var valueRange = fetchValueRangeFromSolution(valueRangeDescriptor, solution, sorter);
+            fromSolution[valueRangeDescriptor.getOrdinal()] =
+                    new ValueRangeItem<>(null, leftValueRange, leftSorter, valueRange, sorter);
+            return valueRange;
+        } else {
+            throw new IllegalStateException(
+                    "Impossible state: the value range (%s) with sorter (%s) does not align with the existing ascending (%s) and descending (%s) sorters."
+                            .formatted(valueRangeDescriptor, sorter, leftSorter, rightSorter));
+        }
+    }
+
+    private Map<Object, Integer> getIndexMapFromSolution(ValueRangeDescriptor<Solution_> valueRangeDescriptor) {
+        getFromSolution(valueRangeDescriptor);
```

Comment:
This is not particularly nice. This is a method that does something, doesn't say what it does, and doesn't return anything.